### PR TITLE
[AKS] api-server RBAC access for watchdog

### PR DESCRIPTION
**Pull request by YanjieGao**
As api-server over AKS need RBAC access, make prometheus & watchdog leverage service acccount of K8s to access api-server for this case.

@xudifsd @ydye ,mv the previous reviewd pr to submit to pai/aks

- Previous reviewed PR:
https://github.com/Microsoft/pai/pull/1601
- Related tracking issue:
https://github.com/Microsoft/pai/issues/1551

### CODE REVIEW

File: deployment/quick-start/services-configuration.yaml.template [hunk 1]
```diff
@@ -41,6 +41,10 @@ cluster:
     # Must be lower case, e.g., regsecret.
     secret-name: pai-secret
 
+  api-server:
```
Comment (ydye) [inline deployment/quick-start/services-configuration.yaml.template]: Don't put this configuration here. It's not a service.

Comment (YanjieGao) [reply deployment/quick-start/services-configuration.yaml.template]: this is flag will be used by prometheus / watchdog / webportal. Do you have suggestions it is better to put this config to which location and file?

Comment (ydye) [reply deployment/quick-start/services-configuration.yaml.template]: kubernetes-configuraiton

File: src/prometheus/deploy/prometheus-configmap.yaml.template [hunk 2]
```diff
@@ -36,6 +36,13 @@ data:
     scrape_configs:
     - job_name: 'node_exporter'
       scrape_interval: {{ prom_info['scrape_interval']|default(30) }}s
+      {% if clusterinfo['api-server']['tls_access'] -%}
+      tls_config:
+        ca_file: /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
```
Comment (fanyangCS) [inline src/prometheus/deploy/prometheus-configmap.yaml.template]: >/var/run/secrets/kubernetes.io/serviceaccount/ca.crt [](start = 17, length = 52)

maybe you should make the path to to keep the ca file and secretes configurable. Remember AKS is not the only k8s deployment you need to support.

Comment (fanyangCS) [reply src/prometheus/deploy/prometheus-configmap.yaml.template]: This change not only affect Prometheus and watchdog. It makes some assumption to the underlying host environment. 

I suggest you to break this into multiple PRs. 1. one makes the CA/secretes path configurable. 2. the other one allows Prometheus to specify the path to CA/Secretes. 3. the final one does the configuration for watchdog.

---
In reply to: [234093754](https://github.com/Microsoft/pai/pull/1710#discussion_r234093754) [](ancestors = 234093754)

Comment (YanjieGao) [reply src/prometheus/deploy/prometheus-configmap.yaml.template]: Will split it and resubmit

Comment (YanjieGao) [reply src/prometheus/deploy/prometheus-configmap.yaml.template]: @ydye how about define the path env and which i can get the path from the env.

**Comment by YanjieGao** [comment]
@xudifsd, @ydye @fanyangCS  
Seperate current PR to 2:
1 watchdog api-server access related is this PR
2 prometheus related is https://github.com/Microsoft/pai/pull/1875

will add unified token path after merge to new cluster object model branch

File: src/watchdog/src/watchdog.py [hunk 3]
```diff
@@ -353,6 +362,9 @@ def main(args):
     parser.add_argument("--log", "-l", help="log dir to store log", default="/datastorage/prometheus")
     parser.add_argument("--interval", "-i", help="interval between two collection", default="30")
     parser.add_argument("--port", "-p", help="port to expose metrics", default="9101")
+    parser.add_argument("--hosts", "-m", help="yaml file path contains host info", default="/etc/watchdog/config.yml")
```
Comment (xudifsd) [inline src/watchdog/src/watchdog.py]: what's this? I didn't see any usage.

Comment (YanjieGao) [reply src/watchdog/src/watchdog.py]: removed

File: src/watchdog/src/watchdog.py [hunk 4]
```diff
@@ -204,11 +204,11 @@ def _map_fn(item):
     map(_map_fn, podsJsonObject["items"])
 
 
-def collect_healthz(gauge, histogram, service_name, address, port, url):
+def collect_healthz(gauge, histogram, service_name, address, port, url, ca_path, headers):
     with histogram.time():
         error = "ok"
         try:
-            error = requests.get("http://{}:{}{}".format(address, port, url)).text
+            error = requests.get("https://{}:{}{}".format(address, port, url), headers = headers, verify = ca_path).text
```
Comment (fanyangCS) [inline src/watchdog/src/watchdog.py]: this assumes the k8s deployment will enable https. and this won't work with existing k8s deployment, right?

File: src/watchdog/src/watchdog.py [hunk 5]
```diff
@@ -217,20 +217,20 @@ def collect_healthz(gauge, histogram, service_name, address, port, url):
...
     # check kubelet
     nodeItems = nodesJsonObject["items"]
 
     for name in nodeItems:
         ip = name["metadata"]["name"]
-
+        
         collect_healthz(k8s_gauge, kubelet_healthz_histogram,
-            "k8s_kubelet", ip, 10255, "/healthz")
+            "k8s_kubelet", "http", ip, 10255, "/healthz", None, None)
```
Comment (fanyangCS) [inline src/watchdog/src/watchdog.py]: where do you detect whether this should be "http" or "https"?

**Comment by fanyangCS** [comment]
neither Jenkins nor any unit test covers this part of code. we need to test on this PR.

File: src/watchdog/src/watchdog.py (hunk 5 above)
Comment (YanjieGao) [reply src/watchdog/src/watchdog.py]: For on-prem existing PAI it use http. 

For AKS I haven't find a method to access this kubelet port. I also suggest this metrics is better to be skipped for pai on AKS, as this responsibility should be taken by AKS and not PAI.

we could add a pending issue for future test.

**Comment by YanjieGao** [comment]
after no issue and approved, will retest it

File: src/watchdog/src/watchdog.py (hunk 5 above)
Comment (fanyangCS) [reply src/watchdog/src/watchdog.py]: in this case you will need to discuss with the team for this decision before submitting a PR.

Comment (YanjieGao) [reply src/watchdog/src/watchdog.py]: Not aware this before, so not mentioned. Currently i haven't make any decision. and keep previous code as much as possible

Comment (YanjieGao) [reply src/watchdog/src/watchdog.py]: @xudifsd , kubelet url access issue created and assigned to you, needs your help to investigate, thanks. 
https://github.com/Microsoft/pai/issues/1878

File: deployment/quick-start/services-configuration.yaml.template (hunk 1 above)
Comment (YanjieGao) [reply deployment/quick-start/services-configuration.yaml.template]: Hi @ydye, after refactor, this config move to kubernetes config .  make it as user configure according to their needs

**Pull request merged**